geo_heatmap.py

```diff
@@ -63,6 +65,35 @@ def loadKMLData(self, file_name):
                 self.coordUpdate(lat_lon)
                 pb.update(i)
 
+    def load_zip_data(self, file_name):
+        """
+        Load google location data from a "takeout-*.zip" file.
+        """
+        from bs4 import BeautifulSoup
+        """
+        <div class="service_name"><h1 class="data-folder-name" data-english-name="LOCATION_HISTORY" data-folder-name="Location History">Location History</h1></div>
+        """
+        zip_file = zipfile.ZipFile(file_name)
+        namelist = zip_file.namelist()
+        (html_path,) = fnmatch.filter(namelist, "Takeout/*.html")
+        with zip_file.open(html_path) as read_file:
+            soup = BeautifulSoup(read_file, "html.parser")
+        (elem,) = soup.select("#service-tile-LOCATION_HISTORY > button > div.service_summary > div > h1[data-english-name=LOCATION_HISTORY]")
+        name = elem['data-folder-name']
+        (data_path,) = fnmatch.filter(namelist, "Takeout/{name}/{name}.*".format(name=name))
+        print("Reading location data file from zip archive: {!r}".format(data_path))
+        if data_path.endswith(".json"):
+            with zip_file.open(data_path) as read_file:
+                self.loadJSONData(read_file)
+        elif data_path.endswith(".kml"):
+            with zip_file.open(data_path) as read_file:
+                self.loadKMLData(read_file)
+        else:
+            raise ValueError(
+                "unsupported extension for {!r}: only .json and .kml supported"
+                .format(file_name)
+            )

```

Comment:
this logic is repetitive with `.run`. but perhaps we can fix this at a later time.